### PR TITLE
MQTT over Websocket

### PR DESCRIPTION
**Pull request by jukkar**
Initial support for running MQTT over Websocket. The mqtt_publisher sample application in this PR, is able to connect to Websocket enabled mosquitto MQTT server.

Note that this PR contains commits from #19025 and all the comments about Websocket or HTTP APIs / samples should be written there.

### CODE REVIEW

File: subsys/net/lib/mqtt/mqtt_transport_websocket.c [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+			       u32_t buflen, bool shall_block)
+{
+	int flags = 0;
+	int ret;
+
+	if (!shall_block) {
+		flags |= MSG_DONTWAIT;
+	}
+
+	ret = recv(client->transport.websocket.sock, data, buflen, flags);
```
Comment (rlubos) [inline subsys/net/lib/mqtt/mqtt_transport_websocket.c]: I think we should be consistent in terms of which API we use (`mqtt_client_websocket_write` calls `websocket_send_msg` while `mqtt_client_websocket_read` calls `recv`).

Personally, I'd stick to the `websocket_*` API, as from my point of view, it should make it easier to get this transport integrated with offloading. After moving the offloading underneath `zsock_*` (hopefully, I could look into this topic soon), both APIs should work, so I won't object if you decide to use socket API, but for now `websocket_*` is the only viable option for offloading.

File: subsys/net/lib/mqtt/mqtt_transport_websocket.c [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+#include <errno.h>
+#include <net/socket.h>
+#include <net/mqtt.h>
+#include <net/websocket.h>
+
+#include "mqtt_os.h"
+#include "mqtt_transport.h"
+
+/**@brief Handles connect request for Websocket transport.
```
Comment (rlubos) [inline subsys/net/lib/mqtt/mqtt_transport_websocket.c]: I have a feeling that we keep duplicating these docs needlessly. This functionality is documented in `mqtt_transport.h`, so it's probably not needed to repat it in specific transport implementations. Personally, I'd remove the docs from other transports as well.

**Review by rlubos — COMMENTED** [review]
Looks pretty good, left a few remarks to consider.

**Review by dbkinder — APPROVED** [review]
doc changes LGTM.

File: subsys/net/lib/mqtt/mqtt_transport.h [hunk 3]
```diff
@@ -95,6 +95,34 @@ int mqtt_transport_read(struct mqtt_client *client, u8_t *data, u32_t buflen,
  */
 int mqtt_transport_disconnect(struct mqtt_client *client);
 
+/* Transport handler functions for TCP socket transport. */
+extern int mqtt_client_tcp_connect(struct mqtt_client *client);
```
[the remark this answers is not in the document]
Comment (jukkar) [reply subsys/net/lib/mqtt/mqtt_transport.h]: Yep, I will remove that.

File: subsys/net/lib/mqtt/mqtt_transport_websocket.c (hunk 1 above)
Comment (jukkar) [reply subsys/net/lib/mqtt/mqtt_transport_websocket.c]: Agreed, I will change the code to use websocket_* API.

Note that the websocket API uses socket interface so the offloading case should be automatically handled.

File: subsys/net/lib/mqtt/mqtt_transport_websocket.c (hunk 2 above)
Comment (jukkar) [reply subsys/net/lib/mqtt/mqtt_transport_websocket.c]: Agreed, I will add a patch to this PR that does that.

**Comment by jukkar** [comment]
Fixed the connectivity issues to mosquitto. The MQTT over Websocket works now in the latest version.

**Review by rlubos — APPROVED** [review]
LGTM

**Comment by jukkar** [comment]
Lets merge #19025 before merging this one. I will keep this PR updated as needed.

**Comment by jukkar** [comment]
Updated according to changes in #19025

**Pull request merged**